service/src/main/kotlin/io/provenance/api/domain/usecase/provenance/tx/permissions/dataAccess/UpdateScopeDataAccess.kt

```diff
@@ -1,48 +1,121 @@
+import tech.figure.objectstore.gateway.client.ClientConfig
+import tech.figure.objectstore.gateway.client.GatewayClient
 
 @Component
 class UpdateScopeDataAccess(
     private val provenanceService: ProvenanceService,
-    private val getSigner: GetSigner
+    private val getSigner: GetSigner,
+    private val provenanceProperties: ProvenanceProperties,
+    private val createGatewayJwt: CreateGatewayJwt,
 ) : AbstractUseCase<UpdateScopeDataAccessRequestWrapper, TxResponse>() {
     override suspend fun execute(args: UpdateScopeDataAccessRequestWrapper): TxResponse {
         val signer = getSigner.execute(GetSignerRequest(args.uuid, args.request.account))
 
-        val messages = args.request.changes.map {
-            when (it.type) {
+        val messages = runActionForChange(
+            args.request.changes,
+            {
+                MsgAddScopeDataAccessRequest.newBuilder()
+                    .setScopeId(MetadataAddress.forScope(args.request.scopeUuid).bytes.toByteString())
+                    .addDataAccess(it.address)
+                    .addAllSigners(listOf(signer.address()))
+                    .build().toAny()
+            },
+            {
+                MsgDeleteScopeDataAccessRequest.newBuilder()
+                    .setScopeId(MetadataAddress.forScope(args.request.scopeUuid).bytes.toByteString())
+                    .addDataAccess(it.address)
+                    .addAllSigners(listOf(signer.address()))
+                    .build().toAny()
+            }
+        ).toTxBody()
+
+        return provenanceService.executeTransaction(args.request.provenanceConfig, messages, signer)
+            .takeIf { !it.isError() }?.let {
+                args.request.objectStoreConfig?.let { osConfig ->
+                    val jwt = createGatewayJwt.execute(CreateGatewayJwtRequest(args.uuid, args.request.account.keyManagementConfig))
+                    GatewayClient(
+                        ClientConfig(
+                            URI.create(osConfig.objectStoreUrl),
+                            provenanceProperties.mainnet
+                        )
+                    ).use { client ->
+                        runActionForChange(
+                            args.request.changes, { change ->
+                            client.grantScopePermission(
+                                MetadataAddress.forScope(args.request.scopeUuid).toString(),
+                                change.address,
+                                jwt
+                            )
+                        },
+                            { change ->
+                                client.revokeScopePermission(
+                                    MetadataAddress.forScope(args.request.scopeUuid).toString(),
+                                    change.address,
+                                    jwt
+                                )
+                            }
+                        )
+                    }
+                }
```

Comment:
Nit: I think it would be slightly cleaner and easier to read this block if the `let` scope was limited to just the creation of `GatewayClient`. Attempting to illustrate this but I may mess up the brackets and/or indentation:
```suggestion
                args.request.objectStoreConfig?.let { osConfig ->
                    val jwt = createGatewayJwt.execute(CreateGatewayJwtRequest(args.uuid, args.request.account.keyManagementConfig))
                    GatewayClient(
                        ClientConfig(
                            URI.create(osConfig.objectStoreUrl),
                            provenanceProperties.mainnet
                        )
                    )
                }.use { client ->
                    runActionForChange(
                        args.request.changes, { change ->
                            client.grantScopePermission(
                                MetadataAddress.forScope(args.request.scopeUuid).toString(),
                                change.address,
                                jwt
                            )
                        },
                            { change ->
                                client.revokeScopePermission(
                                    MetadataAddress.forScope(args.request.scopeUuid).toString(),
                                    change.address,
                                    jwt
                                )
                            }
                        )
                    }
```